pkg/operator/controller/ingress_status.go

```diff
@@ -101,3 +126,156 @@ func ingressStatusesEqual(a, b operatorv1.IngressControllerStatus) bool {
 
 	return true
 }
+
+func isProvisioned(service *corev1.Service) bool {
+	ingresses := service.Status.LoadBalancer.Ingress
+	return len(ingresses) > 0 && (len(ingresses[0].Hostname) > 0 || len(ingresses[0].IP) > 0)
+}
+
+func isPending(service *corev1.Service) bool {
+	return !isProvisioned(service)
+}
+
+func getServiceOwnerIfMatches(service *corev1.Service, matches func(*corev1.Service) bool) []string {
+	if !matches(service) {
+		return []string{}
+	}
+	controller, ok := service.Labels[manifests.OwningIngressControllerLabel]
+	if !ok {
+		return []string{}
+	}
+	return []string{controller}
+}
+
+func indexLoadBalancerControllerByName(obj runtime.Object) []string {
+	c, ok := obj.(*operatorv1.IngressController)
+	if !ok {
+		return []string{}
+	}
+	if c.Status.EndpointPublishingStrategy != nil &&
+		c.Status.EndpointPublishingStrategy.Type == operatorv1.LoadBalancerServiceStrategyType {
+		return []string{c.Name}
+	}
+	return []string{}
+}
+
+func indexProvisionedLoadBalancerServiceByOwner(obj runtime.Object) []string {
+	service, ok := obj.(*corev1.Service)
+	if !ok {
+		return []string{}
+	}
+	if service.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return []string{}
+	}
+	return getServiceOwnerIfMatches(service, isProvisioned)
+}
+
+func indexPendingLoadBalancerServiceByOwner(obj runtime.Object) []string {
+	service, ok := obj.(*corev1.Service)
+	if !ok {
+		return []string{}
+	}
+	if service.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		return []string{}
+	}
+	return getServiceOwnerIfMatches(service, isPending)
+}
+
+// ingressStatusCache knows how to compute status for ingress controllers by
+// querying indexes caches.
+type ingressStatusCache struct {
+	// contains returns true if there are >0 matches on value for the named index
+	// for the given kind.
+	//
+	// listKind must be the List type for the object being indexed.
+	contains func(listKind runtime.Object, name, value string) bool
+}
+
+// serviceIndexers are all the Service indexes supported by the cache.
+var serviceIndexers = map[string]client.IndexerFunc{
+	"pending-for":     indexPendingLoadBalancerServiceByOwner,
+	"provisioned-for": indexProvisionedLoadBalancerServiceByOwner,
+}
+
+// controllerIndexers are all the IngressController indexes supported by the
+// cache.
+var controllerIndexers = map[string]client.IndexerFunc{
+	"wants-load-balancer": indexLoadBalancerControllerByName,
+}
+
+// cacheContains is a contains fuction which knows how to query a cache.Cache.
```

Comment:
"fuction"→"function".

The phrase "is a contains function" implies that this function implements some "contains" interface, but its signature doesn't quite match the type of `contains` in `ingressStatusCache`.  Maybe it would be more helpful to say describe the function more abstractly; e.g., "cacheContains returns membership of the given value in the named index for the specified kind using the given cache."

Come to think of it, the `list` parameter is odd—to the caller, its purpose is to convey a resource kind, but the callee uses it to store a list of values of that kind, which forces the parameter to be the type for a list of resources of the kind.  It would be a more obvious API if callers instead could specify the type for the kind.  Alternatively, we could drop the `list` parameter and look up the type from the index using a map.